File: pkg/cli/cmd/provider.go
/*
Copyright © 2026 masteryyh <yyh991013@163.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package cmd

import (
	"fmt"
	"os"
	"strings"

	"github.com/masteryyh/agenty/pkg/models"
	"github.com/pterm/pterm"
	"github.com/spf13/cobra"
	"golang.org/x/term"
)

var providerCmd = &cobra.Command{
	Use:   "provider",
	Short: "Manage model providers",
	Long:  `Create, list, update and delete model providers`,
}

var providerListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all providers",
	RunE: func(cmd *cobra.Command, args []string) error {
		c := GetClient()

		page, _ := cmd.Flags().GetInt("page")
		pageSize, _ := cmd.Flags().GetInt("page-size")

		result, err := c.ListProviders(page, pageSize)
		if err != nil {
			return err
		}

		if len(result.Data) == 0 {
			pterm.Warning.Println("No providers found")
			return nil
		}

		tableData := pterm.TableData{
			{"Name", "Type", "Base URL", "API Key"},
		}
		for _, p := range result.Data {
			tableData = append(tableData, []string{p.Name, string(p.Type), p.BaseURL, p.APIKeyCensored})
		}
		pterm.DefaultTable.WithHasHeader().WithData(tableData).Render()
		pterm.Info.Printf("Total: %d, Page: %d/%d\n", result.Total, result.Page, (result.Total+int64(result.PageSize)-1)/int64(result.PageSize))
		return nil
	},
}

var providerTypeOptions = []string{"openai", "anthropic", "gemini", "kimi"}

var providerDefaultBaseURLs = map[string]string{
	"openai":    "https://api.openai.com/v1",
	"anthropic": "https://api.anthropic.com",
	"gemini":    "https://generativelanguage.googleapis.com",
	"kimi":      "https://api.moonshot.cn/v1",
}

func readHiddenInput(prompt string) (string, error) {
	pterm.DefaultInteractiveTextInput.TextStyle.Print(prompt + ": ")
	key, err := term.ReadPassword(int(os.Stdin.Fd()))
	fmt.Println()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(key)), nil
}

func providerLabel(p models.ModelProviderDto) string {
	return fmt.Sprintf("%s (%s)", p.Name, string(p.Type))
}

var providerCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a new provider",
	RunE: func(cmd *cobra.Command, args []string) error {
		if !term.IsTerminal(int(os.Stdin.Fd())) {
			return fmt.Errorf("must be run in an interactive terminal")
		}

		c := GetClient()

		name, err := pterm.DefaultInteractiveTextInput.Show("Provider name")
		if err != nil {
			return err
		}
		name = strings.TrimSpace(name)
		if name == "" {
			return fmt.Errorf("name cannot be empty")
		}

		selectedType, err := pterm.DefaultInteractiveSelect.
			WithOptions(providerTypeOptions).
			Show("Provider type")
		if err != nil {
			return err
		}

		baseURL, err := pterm.DefaultInteractiveTextInput.
			WithDefaultValue(providerDefaultBaseURLs[selectedType]).
			Show("Base URL")
		if err != nil {
			return err
		}
		baseURL = strings.TrimSpace(baseURL)
		if baseURL == "" {
			return fmt.Errorf("base URL cannot be empty")
		}

		apiKey, err := readHiddenInput("API key")
		if err != nil {
			return err
		}

		provider, err := c.CreateProvider(&models.CreateModelProviderDto{
			Name:    name,
			Type:    models.APIType(selectedType),
			BaseURL: baseURL,
			APIKey:  apiKey,
		})
		if err != nil {
			return err
		}
		pterm.Success.Printf("Provider created: %s\n", provider.Name)
		return nil
	},
}

var providerUpdateCmd = &cobra.Command{
	Use:   "update",
	Short: "Update a provider",
	RunE: func(cmd *cobra.Command, args []string) error {
		if !term.IsTerminal(int(os.Stdin.Fd())) {
			return fmt.Errorf("must be run in an interactive terminal")
		}

		c := GetClient()

		result, err := c.ListProviders(1, 100)
		if err != nil {
			return fmt.Errorf("failed to list providers: %w", err)
		}
		if len(result.Data) == 0 {
			pterm.Warning.Println("No providers found")
			return nil
		}

		options := make([]string, len(result.Data))
		for i, p := range result.Data {
			options[i] = providerLabel(p)
		}
		selected, err := pterm.DefaultInteractiveSelect.WithOptions(options).Show("Select provider to update")
		if err != nil {
			return err
		}
		var target models.ModelProviderDto
		for _, p := range result.Data {
			if providerLabel(p) == selected {
				target = p
				break
			}
		}

		pterm.Info.Println("Press Enter to keep the current value")

		newName, err := pterm.DefaultInteractiveTextInput.WithDefaultValue(target.Name).Show("Name")
		if err != nil {
			return err
		}
		newName = strings.TrimSpace(newName)
		if newName == "" {
			newName = target.Name
		}

		newType, err := pterm.DefaultInteractiveSelect.
			WithOptions(providerTypeOptions).
			WithDefaultOption(string(target.Type)).
			Show("Type")
		if err != nil {
			return err
		}

		newBaseURL, err := pterm.DefaultInteractiveTextInput.WithDefaultValue(target.BaseURL).Show("Base URL")
		if err != nil {
			return err
		}
		newBaseURL = strings.TrimSpace(newBaseURL)
		if newBaseURL == "" {
			newBaseURL = target.BaseURL
		}

		newAPIKey, err := readHiddenInput("API key (leave empty to keep unchanged)")
		if err != nil {
			return err
		}
		newAPIKey = strings.TrimSpace(newAPIKey)

		if target.Name == newName && string(target.Type) == newType && target.BaseURL == newBaseURL && newAPIKey == "" {
			pterm.Info.Println("No changes detected, skipping update")
			return nil
		}

		updated, err := c.UpdateProvider(target.ID, &models.UpdateModelProviderDto{
			Name:    newName,
			Type:    models.APIType(newType),
			BaseURL: newBaseURL,
			APIKey:  newAPIKey,
		})
		if err != nil {
			return err
		}
		pterm.Success.Printf("Provider updated: %s\n", updated.Name)
		return nil
	},
}

var providerDeleteCmd = &cobra.Command{
	Use:   "delete",
	Short: "Delete a provider",
	RunE: func(cmd *cobra.Command, args []string) error {
		if !term.IsTerminal(int(os.Stdin.Fd())) {
			return fmt.Errorf("must be run in an interactive terminal")
		}

		c := GetClient()

		result, err := c.ListProviders(1, 100)
		if err != nil {
			return fmt.Errorf("failed to list providers: %w", err)
		}
		if len(result.Data) == 0 {
			pterm.Warning.Println("No providers found")
			return nil
		}

		options := make([]string, len(result.Data))
		for i, p := range result.Data {
			options[i] = providerLabel(p)
		}
		selected, err := pterm.DefaultInteractiveSelect.WithOptions(options).Show("Select provider to delete")
		if err != nil {
			return err
		}
		var target models.ModelProviderDto
		for _, p := range result.Data {
			if providerLabel(p) == selected {
				target = p
				break
			}
		}

		force, err := pterm.DefaultInteractiveConfirm.Show(fmt.Sprintf("Also delete all models under '%s'?", target.Name))
		if err != nil {
			return err
		}

		confirm, err := pterm.DefaultInteractiveConfirm.Show(fmt.Sprintf("Delete provider '%s'?", target.Name))
		if err != nil {
			return err
		}
		if !confirm {
			pterm.Info.Println("Cancelled")
			return nil
		}

		if err := c.DeleteProvider(target.ID, force); err != nil {
			return err
		}
		pterm.Success.Printf("Provider deleted: %s\n", target.Name)
		return nil
	},
}

func init() {
	rootCmd.AddCommand(providerCmd)

	providerCmd.AddCommand(providerListCmd)
	providerListCmd.Flags().Int("page", 1, "Page number")
	providerListCmd.Flags().Int("page-size", 10, "Page size")

	providerCmd.AddCommand(providerCreateCmd)
	providerCmd.AddCommand(providerUpdateCmd)
	providerCmd.AddCommand(providerDeleteCmd)
}

File: agenty-cli/src/cli/provider.ts
/*
Copyright © 2026 masteryyh <yyh991013@163.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import type { AgentyClient } from "@/api/client";
import type { UpdateModelProviderDto } from "@/api/types";
import {
	action,
	CliError,
	flag,
	hasFlag,
	outputFields,
	outputTable,
	pageOptions,
	render,
	requireFlag,
	requirePositionals,
	resolveProvider,
	secret,
	type ParsedArgs,
} from "./utils";

const subcommandHandlers: Record<
	string,
	(client: AgentyClient, args: ParsedArgs) => Promise<void>
> = {
	list: handleListProvider,
	get: handleGetProvider,
	add: handleAddProvider,
	update: handleUpdateProvider,
	remove: handleRemoveProvider,
};

export async function handleProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	const subcommand = args.positionals[1];
	const handler = subcommand ? subcommandHandlers[subcommand] : undefined;
	if (!handler) {
		throw new CliError("usage: provider <list|get|add|update|remove>");
	}
	await handler(client, args);
}

async function handleListProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	requirePositionals(args, 2, "provider list");
	const { page, pageSize } = pageOptions(args);
	const result = await client.listProvidersPage(page, pageSize);

	render(args, result, () =>
		result.data.length === 0
			? process.stdout.write("No providers.\n")
			: outputTable(
					["ID", "Name", "Type", "Base URL", "API Key"],
					result.data.map((provider) => [
						provider.id,
						provider.name,
						provider.type,
						provider.baseUrl,
						provider.apiKeyCensored,
					]),
				),
	);
}

async function handleGetProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	const [, , reference] = requirePositionals(
		args,
		3,
		"provider get <name-or-id>",
	);
	const provider = await resolveProvider(client, reference);
	render(args, provider, () =>
		outputFields([
			["ID", provider.id],
			["Name", provider.name],
			["Type", provider.type],
			["Base URL", provider.baseUrl],
			[
				"Bailian multimodal embedding base URL",
				provider.bailianMultiModalEmbeddingBaseUrl ?? "",
			],
			["API Key", provider.apiKeyCensored],
			["Preset", String(provider.isPreset)],
		]),
	);
}

async function handleAddProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	const [, , name] = requirePositionals(
		args,
		3,
		"provider add <name> --type <type> --base-url <url> --api-key <key>",
	);
	const apiKey = secret(args, "api-key", "api-key-env", "provider API key");
	if (!apiKey) {
		throw new CliError("--api-key or --api-key-env is required");
	}
	const created = await client.createProvider({
		name,
		type: requireFlag(args, "type"),
		baseUrl: requireFlag(args, "base-url"),
		apiKey,
		...(hasFlag(args, "bailian-multimodal-embedding-base-url")
			? {
					bailianMultiModalEmbeddingBaseUrl: requireFlag(
						args,
						"bailian-multimodal-embedding-base-url",
					),
				}
			: {}),
	});
	action(args, created, `Provider added: ${created.name}`);
}

async function handleUpdateProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	const [, , reference] = requirePositionals(
		args,
		3,
		"provider update <name-or-id> [options]",
	);
	const current = await resolveProvider(client, reference);
	const dto: UpdateModelProviderDto = {};
	for (const [flagName, field] of [
		["name", "name"],
		["type", "type"],
		["base-url", "baseUrl"],
		[
			"bailian-multimodal-embedding-base-url",
			"bailianMultiModalEmbeddingBaseUrl",
		],
	] as const) {
		if (hasFlag(args, flagName)) {
			dto[field] = requireFlag(args, flagName);
		}
	}
	const apiKey = secret(args, "api-key", "api-key-env", "provider API key");
	if (apiKey) {
		dto.apiKey = apiKey;
	}
	if (Object.keys(dto).length === 0) {
		throw new CliError("no changes specified");
	}

	const updated = await client.updateProvider(current.id, dto);
	action(args, updated, `Provider updated: ${updated.name}`);
}

async function handleRemoveProvider(
	client: AgentyClient,
	args: ParsedArgs,
): Promise<void> {
	const [, , reference] = requirePositionals(
		args,
		3,
		"provider remove <name-or-id> --yes",
	);
	if (!hasFlag(args, "yes")) {
		throw new CliError("use --yes to remove a provider non-interactively");
	}
	const current = await resolveProvider(client, reference);
	await client.deleteProvider(current.id);
	action(
		args,
		{ id: current.id, name: current.name, deleted: true },
		`Provider removed: ${current.name}`,
	);
}
